Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/KeepAliveServiceEndpointDecorator.java

```diff
@@ -0,0 +1,135 @@
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package gov.hhs.fha.nhinc.messaging.service.decorator;
+
+import gov.hhs.fha.nhinc.common.nhinccommon.AssertionType;
+import gov.hhs.fha.nhinc.messaging.service.ServiceEndpoint;
+import gov.hhs.fha.nhinc.nhinclib.NhincConstants;
+import gov.hhs.fha.nhinc.nhinclib.NullChecker;
+import gov.hhs.fha.nhinc.properties.PropertyAccessException;
+import gov.hhs.fha.nhinc.properties.PropertyAccessor;
+import javax.xml.ws.BindingProvider;
+import org.apache.cxf.transports.http.configuration.ConnectionType;
+import org.apache.cxf.transports.http.configuration.HTTPClientPolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class will allow users to turn on/off KeepAlive feature in http header by detecing keepalive tag inside
+ * assertion block. This also allow users to turn on/off KeepAlive inside gateway.properties. However, the assertion
+ * element will take precedence over the gateway.properties. All values are in seconds. To turn off, either remove
+ * keepAlive elements or using integer < 1
+ *
+ * @author mpnguyen
+ */
+public class KeepAliveServiceEndpointDecorator<T> extends ServiceEndpointDecorator<T> {
+    private static final String KEEP_ALIVE = "keepAliveDuration";
+    private final AssertionType assertion;
+    private final String serviceName;
+    private static final Logger LOG = LoggerFactory.getLogger(KeepAliveServiceEndpointDecorator.class);
+    /**
+     * @param decoratedEndpoint
+     */
+    public KeepAliveServiceEndpointDecorator(final ServiceEndpoint<T> decoratedEndpoint, final AssertionType assertion,
+            final String serviceName) {
+        super(decoratedEndpoint);
+        this.assertion = assertion;
+        this.serviceName = serviceName;
+        LOG.debug("Initialize KeepAliveServiceEndpointDecorator {}", serviceName);
+    }
+    @Override
+    public void configure() {
+        super.configure();
+        final HTTPClientPolicy httpClientPolicy = getHTTPClientPolicy();
+        final BindingProvider bindingProvider = (BindingProvider) getPort();
+        // reset keepAlive duration
+        bindingProvider.getRequestContext().remove(NhincConstants.KEEPALIVE_DURATION);
+        //keepalive from assertion block has precedence over the default properties.
+        // KeepAlive Null means user is not aware of this feature. Will turn on if default indicates so.
+        // KeepAlive < 0 mean user explicitly wants to turn off regardless default value
+        // KeepAlive > 0 mean user explicitly wants to turn on with indicated value.
+        final Integer keepAliveAssertion = assertion.getKeepAlive();
+        final StringBuilder logMsgBuilder = new StringBuilder("KeepAlive statement");
+        logMsgBuilder.append(" for ");
+        logMsgBuilder.append(this.serviceName);
+        logMsgBuilder.append(" service ");
+        if (keepAliveAssertion == null) {// msg comes empty
```

Comment:
Stylistic suggestion -- usually, comments should not be on the same line as code.